### PR TITLE
Add firewall to limit network access for runners

### PR DESCRIPTION
**Pull request by weiiwang01**
Add a charm configuration `allowlist` to configure runners to only access necessary ports and IP addresses, minimizing the potential surface for attacks.

### CODE REVIEW

File: templates/github-runner-firewall.j2 [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        iifname "tap*" ip daddr { {{ host_ip }}, 255.255.255.255 } udp dport 67 counter accept
+        iifname "tap*" ip daddr {{ host_ip }} tcp dport 8080 counter accept
+
+{% for entry in allowlist %}
+        iifname "tap*" ip daddr {{ entry.ip_range }} {{ "udp" if entry.is_udp else "tcp" }} dport {{ entry.port_range }} counter accept
+{% endfor %}
+
+        iifname "tap*" counter reject with icmpx type admin-prohibited
+    }
+}
```
Comment (jdkandersson) [inline templates/github-runner-firewall.j2:21]: Should this include a final EOL?

**Comment by jdkandersson** [comment]
Thanks for putting this together!

File: templates/github-runner-firewall.j2 (hunk 1 above)
Comment (weiiwang01) [reply templates/github-runner-firewall.j2]: Added, thanks!

File: src/firewall.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                raise_format_error()
+            port_start, port_end = port_range_parts
+            check_valid_port(port_start)
+            check_valid_port(port_end)
+        else:
+            check_valid_port(port_range)
+        return cls(ip_range=ip_range, port_range=port_range, is_udp=is_udp)
+
+
+class Firewall:  # pylint: disable=too-few-public-methods
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/firewall.py:78]: I think the name of the check, `too-few-public-methods`, is self-explanatory enough in this situation.

**Review by jdkandersson — DISMISSED** [review]

File: src/firewall.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                ipaddress.IPv4Network(network)
+            except ValueError:
+                raise_format_error()
+
+        def check_valid_port(port: str):
+            """Check if the given string is a valid port number."""
+            if not port.isdigit() or int(port) < 1 or int(port) > 65535:
+                raise_format_error()
+
+        parts = entry.split(":")
```
Comment (arturo-seijas) [inline src/firewall.py:58]: Is something like 0.0.0.0/0:1-65535:udp expected to fail?

File: tests/unit/test_firewall.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    ),
+    pytest.param("10.0.0.1/8:11-22:udp", None),
+    pytest.param("a.b.c.d:123", None),
+    pytest.param("", None),
+    pytest.param("1.2.3.4/a", None),
+)
+
+
+@pytest.mark.parametrize("entry,excepted_firewall_entry", TESTING_FIREWALL_ENTRIES)
+def test_parse_firewall_entry(entry: str, excepted_firewall_entry: typing.Optional[FirewallEntry]):
```
Comment (arturo-seijas) [inline tests/unit/test_firewall.py:28]: s/excepted_firewall_entry/expected_firewall_entry?

File: src/charm.py [hunk 5]
```diff
@@ -549,7 +563,9 @@ def _install_deps(self) -> None:
             env["NO_PROXY"] = self.proxies["no_proxy"]
             env["no_proxy"] = self.proxies["no_proxy"]
 
-        execute_command(["/usr/bin/apt-get", "install", "-qy", "gunicorn", "python3-pip"])
+        execute_command(
+            ["/usr/bin/apt-get", "install", "-qy", "gunicorn", "python3-pip", "nftables"]
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py:567]: It is used by repo-policy-compliance.

File: tests/unit/test_firewall.py (hunk 4 above)
Comment (weiiwang01) [reply tests/unit/test_firewall.py]: Fixed, thanks!

File: src/charm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply src/charm.py:567]: Can you add a comment in the code mentioning this? It isn't very obvious from the outside :)

[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py:568]: 👍 added

[the remark this answers is not in the document]
Comment (arturo-seijas) [reply src/charm.py:568]: Thanks!

**Review by arturo-seijas — DISMISSED** [review]
LGTM

**Review by arturo-seijas — DISMISSED** [review]
LGTM

File: src/charm.py [hunk 6]
```diff
@@ -577,8 +583,8 @@ def _install_deps(self) -> None:
         execute_command(["/usr/bin/snap", "refresh", "lxd", "--channel=latest/stable"])
         execute_command(["/snap/bin/lxd", "waitready"])
         execute_command(["/snap/bin/lxd", "init", "--auto"])
-        execute_command(["/usr/bin/chmod", "a+wr", "/var/snap/lxd/common/lxd/unix.socket"])
         execute_command(["/snap/bin/lxc", "network", "set", "lxdbr0", "ipv6.address", "none"])
+        execute_command(["/usr/bin/chmod", "a+wr", "/var/snap/lxd/common/lxd/unix.socket"])
```
Comment (amandahla) [inline src/charm.py:587]: I believe it's out of the scope of this PR but why is this needed?

Comment (weiiwang01) [reply src/charm.py:587]: I guess this is to call lxc command without sudo, but since the charm runs as root, I don't think this is technically necessary.

Comment (weiiwang01) [reply src/charm.py]: I just removed that line and it seems to be fine.

**Review by amandahla — DISMISSED** [review]
LGTM

File: config.yaml [hunk 7]
```diff
@@ -58,3 +58,10 @@ options:
     description: >
       When set to 'insecure', the charm test mode is activated, which may deactivate some security
       hardening measures.
+  allowlist:
+    type: string
+    default: "0.0.0.0/0:1-65535, 0.0.0.0/0:1-65535:udp"
+    description: >
+      A list of IPv4 networks and tcp/udp ports that runners can access. Note that ICMP traffic 
+      and IPv6 traffic are always disallowed. Additionally, the runner will always have
+      access to essential services such as DHCP and DNS regardless of the allowlist configuration.
```
Comment (gtrkiller) [inline config.yaml:67]: What is the logic behind allowing everything in IPv4 by default from a security perspective?

Comment (weiiwang01) [reply config.yaml:67]: For the default value, we only have two options: allow all or allow none. But when the charm is set to allow none, the GitHub runner can't even start, so I opt to chose allow all as the default value.

Comment (gtrkiller) [reply config.yaml]: Makes sense, thanks!

File: .github/workflows/e2e_test.yaml [hunk 8]
```diff
@@ -190,3 +198,7 @@ jobs:
       # Test program installed by pip. The directory `~/.local/bin` need to be added to PATH.
       - name: test check-jsonschema
         run: check-jsonschema --version
+      - name: Test Firewall
+        run: |
+          HOST_IP=$(ip route | grep default | cut -f 3 -d" ")
+          [ $((ping $HOST_IP -c 5 || :) | grep "Destination Port Unreachable" | wc -l) -eq 5 ]
```
Comment (jdkandersson) [inline .github/workflows/e2e_test.yaml:204]: Final newline?

File: src/firewall.py [hunk 9]
```diff
@@ -38,87 +35,118 @@ def decode(cls, entry: str) -> "FirewallEntry":
...
+            {
+                "action": "reject",
+                "destination": host_ip,
+                "protocol": "icmp4",
+                "state": "enabled",
+            },
+            {
+                "action": "reject",
+                "destination": "::/0",
+                "state": "enabled",
```
Comment (jdkandersson) [inline src/firewall.py:128]: What is the effect of this rule please?

Comment (weiiwang01) [reply src/firewall.py:128]: Reject all IPv6 traffics, since we don't provision a IPv6 network to the runner right now.

File: .github/workflows/e2e_test.yaml (hunk 8 above)
Comment (weiiwang01) [reply .github/workflows/e2e_test.yaml]: Updated, thanks!

File: src/firewall.py (hunk 9 above)
Comment (jdkandersson) [reply src/firewall.py:128]: This feels a little Canonical specific, should that go to the denylist configuration rather than in our code? If someone else installs the charm on an IPv6 network I'm guessing it wouldn't work with this configuration?

Comment (weiiwang01) [reply src/firewall.py:128]: The LXD network inside the runner has IPv6 disabled as well, so I suspect it will work even with this rule removed.

Comment (jdkandersson) [reply src/firewall.py:128]: Ah ok

**Review by jdkandersson — APPROVED** [review]

File: src/firewall.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    """
+
+    ip_range: str
+
+    @classmethod
+    def decode(cls, entry: str) -> "FirewallEntry":
+        """Decode a firewall entry from a string.
+
+        Args:
+            entry: The firewall entry string, e.g. '192.168.0.1:80' or '192.168.0.0/24:80-90:udp'.
```
Comment (gregory-schiano) [inline src/firewall.py:30]: Shouldn't the description be adapted ? Only CIDR notation is supported

File: src/firewall.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                "protocol": "icmp4",
+                "state": "enabled",
+            },
+            {
+                "action": "reject",
+                "destination": "::/0",
+                "state": "enabled",
+            },
+        ]
+        host_network = ipaddress.IPv4Network(host_ip)
```
Comment (gregory-schiano) [inline src/firewall.py:131]: The variable name could be confusing, it's not the real host_network, but the host_ip represented as a /32 CIDR notation to be able to use `address_exclude` that requires an IPv4Network object

File: .github/workflows/e2e_test.yaml [hunk 12]
```diff
@@ -109,6 +112,7 @@ jobs:
             --config path=${{ secrets.E2E_TESTING_REPO }} \
             --config token=${{ secrets.E2E_TESTING_TOKEN }} \
             --config virtual-machines=1 \
+            --config denylist=10.0.0.0/8 \
```
Comment (gregory-schiano) [inline .github/workflows/e2e_test.yaml:115]: Does it cover host_ip which is included in the denylist range ?

File: .github/workflows/e2e_test.yaml [hunk 13]
```diff
@@ -100,6 +100,9 @@ jobs:
         with:
           name: dangerous-test-only-github-runner_ubuntu-22.04-amd64.charm
 
+      - name: Enable br_netfilter
+        run: sudo modprobe br_netfilter
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply .github/workflows/e2e_test.yaml:104]: Interesting, the `br_netfilter` is not necessary for the ACL firewall. It's only required by `security.ipv6_filtering=true`

File: src/firewall.py (hunk 10 above)
Comment (weiiwang01) [reply src/firewall.py:30]: Fixed, thanks!

File: src/firewall.py (hunk 11 above)
Comment (weiiwang01) [reply src/firewall.py:131]: Yes, the variable is a bad idea, I removed the variable and inlined the IPv4Network casting.

File: .github/workflows/e2e_test.yaml (hunk 12 above)
Comment (weiiwang01) [reply .github/workflows/e2e_test.yaml:115]: Yes, this is intended to test the capability of the charm to resolve firewall conflicts.

File: .github/workflows/e2e_test.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (gregory-schiano) [reply .github/workflows/e2e_test.yaml:104]: Interesting... curious to understand how they manage the ACL then. Anyway unrelated to this PR, closing the comment

**Review by gregory-schiano — APPROVED** [review]
LGTM
